File: NeetCode/1-Arrays & Hashing/7-Product of Array Except Self/238-Product-of-Array-Except-Self.ts
function productExceptSelf(nums: number[]): number[] {
    const len=nums.length;
    const result=new Array(len).fill(1);
    // [1,2,3,4]
    // [1,1,2,6]
    for(let i=1;i<len;i++)result[i]=nums[i-1]*result[i-1];
    let postfix=1;
    for(let i=len-1;i>=0;i--){
        result[i]=postfix*result[i];
        postfix*=nums[i];
    }
    return result;
};

File: NeetCode/1-Arrays & Hashing/7-Product of Array Except Self/238-Product-of-Array-Except-Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        const int len=nums.size();
        vector<int> prefix(len,1);
        vector<int>postfix(len,1);

        for(int i=1;i<len;i++)prefix[i]=prefix[i-1]*nums[i-1];
        for(int i=len-2;i>=0;i--)postfix[i]=postfix[i+1]*nums[i+1];
        vector<int>result(len);
        for(int i=0;i<len;i++)result[i]=prefix[i]*postfix[i];
        return  result;

    }
};

